### PR TITLE
Fixed logic in the Compression/Decompression gem registration

### LINKED ISSUE #12030: The Compression Gem Compression and Decompression Factory incorrectly handles registration
**Describe the bug**
The Compression Gem contains two factory classes of [CompressionFactory](https://github.com/o3de/o3de/blob/development/Gems/Compression/Code/Source/Tools/CompressionFactoryImpl.h#L31-L38) and [DecompressionFactory](https://github.com/o3de/o3de/blob/development/Gems/Compression/Code/Source/Clients/DecompressionFactoryImpl.h#L30-L38) which is responsible for registration compression and decompression interfaces respectively.

The problem is that it uses some complex logic of having a compression interfaces vector and then a second vector which maps compression algorithm IDs to indices within that vector.
It does this to try to allow binary searching for the registered compression interfaces while supporting cache friendliness.
However the logic is flawed and will cause an invalid memory access should a compression/decompression interface because the second compression algorithm ID to index vector isn't updated when a compression interface is removed

**Assets required**

**Steps to reproduce**
1. Enable the [Compression](https://github.com/o3de/o3de/blob/development/Gems/Compression/gem.json) Gem.
2. Implement a test within the Compression Gem which registers several instances of a Mock CompressionInterface interface with the CompressionFactory
3. Attempt to unregister those Mock Compression Interfaces

**Expected behavior**
Unregistration of the Compressions Interfaces with the factory should succeed

**Actual behavior**
Registration fails due to the compression algorithm ID to index vector not adjusting all the indices that point to elements after the removed elements in the compression interfaces vector to be accurate

**Screenshots/Video**

**Found in Branch**
developer

**Desktop/Device (please complete the following information):**
All

**Additional context**
The approach done in the [EventLoggerFactoryImpl.h/.cpp](https://github.com/o3de/o3de/blob/660b83b1a67e1a45c3b285c0ac52f5a1fb838672/Code/Framework/AzCore/AzCore/Metrics/EventLoggerFactoryImpl.h#L59-L76) for registering interfaces should be used instead.

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Renamed the CompressionFactory and DecompressionFactory classes to CompressionRegistrar and DecompressionRegistrar

fixes #12030

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

### CODE REVIEW

**Review by byrcolin — APPROVED** [review]

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 1]
```diff
@@ -70,28 +69,44 @@ namespace Compression
         //! Invokes the supplied visitor for each register compression interface that is non-nullptr
         virtual void VisitCompressionInterfaces(const VisitCompressionInterfaceCallback&) const = 0;
 
-        //! Registers a compression interface with the compression factory.
+        //! Registers compression interface and takes ownership of it if registration is successful
+        //! @param algorithmId Unique id to associate with compression interface
+        //! @param compressionInterface compression interface to register
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:74]: nit: This parameter name omitted in the function signature below

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 2]
```diff
@@ -70,28 +69,44 @@ namespace Compression
...
         //! If a compression interface with a CompressionAlgorithmId is registered that
         //! matches the input compression interface, then registration does not occur
-        //! and the unique_ptr refererence is unmodified and can be re-used by the caller
         //!
-        //! @param compressionInterface unique pointer to compression interface to register
-        //! @return true if the ICompressionInterface was successfully registered, otherwise false
-        virtual bool RegisterCompressionInterface(AZStd::unique_ptr<ICompressionInterface>&&) = 0;
+        //! Registers compression interface, but does not take ownership of it
+        //! @param algorithmId Unique id to associate with compression interface
+        //! @param compressionInterface compression interface to register
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:87]: nit: This parameter name omitted in the function signature below

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 3]
```diff
@@ -70,28 +69,44 @@ namespace Compression
...
         virtual void VisitCompressionInterfaces(const VisitCompressionInterfaceCallback&) const = 0;
 
-        //! Registers a compression interface with the compression factory.
+        //! Registers compression interface and takes ownership of it if registration is successful
+        //! @param algorithmId Unique id to associate with compression interface
+        //! @param compressionInterface compression interface to register
+        //! @return Success outcome if the compression interface was successfully registered
+        //! Otherwise, a failure outcome with the compression interface is forward back to the caller
+        virtual AZ::Outcome<void, AZStd::unique_ptr<ICompressionInterface>> RegisterCompressionInterface(
+            CompressionAlgorithmId algorithmId, AZStd::unique_ptr<ICompressionInterface>) = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:78]:
```suggestion
            CompressionAlgorithmId algorithmId, AZStd::unique_ptr<ICompressionInterface> compressionInterface) = 0;
```

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 4]
```diff
@@ -70,28 +69,44 @@ namespace Compression
...
-        //! and the unique_ptr refererence is unmodified and can be re-used by the caller
         //!
-        //! @param compressionInterface unique pointer to compression interface to register
-        //! @return true if the ICompressionInterface was successfully registered, otherwise false
-        virtual bool RegisterCompressionInterface(AZStd::unique_ptr<ICompressionInterface>&&) = 0;
+        //! Registers compression interface, but does not take ownership of it
+        //! @param algorithmId Unique id to associate with compression interface
+        //! @param compressionInterface compression interface to register
+        //! @return true if the ICompressionInterface was successfully registered
+        virtual bool RegisterCompressionInterface(CompressionAlgorithmId algorithmId, ICompressionInterface&) = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:89]:
```suggestion
        virtual bool RegisterCompressionInterface(CompressionAlgorithmId algorithmId, ICompressionInterface& compressionInterface ) = 0;
```

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 5]
```diff
@@ -70,28 +69,44 @@ namespace Compression
...
-        //! @return pointer to the Compression Interface or nullptr if not found
-        virtual ICompressionInterface* FindCompressionInterface(CompressionAlgorithmId) const = 0;
+        //! @param compressionAlgorithmId unique Id that identifies the compression interface
+        //! @return true if the unregistration is successful
+        virtual bool UnregisterCompressionInterface(CompressionAlgorithmId algorithmId) = 0;
+
+        //! Queries the compression interface with the compression algorithmd Id
+        //! @param compressionAlgorithmId unique Id of compression interface to query
+        //! @return pointer to the compression interface or nullptr if not found
+        [[nodiscard]] virtual ICompressionInterface* FindCompressionInterface(CompressionAlgorithmId) const = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:100]: Should match the doxygen comment
```suggestion
        [[nodiscard]] virtual ICompressionInterface* FindCompressionInterface(CompressionAlgorithmId compressionAlgorithmId ) const = 0;
```

File: Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h [hunk 6]
```diff
@@ -70,28 +69,44 @@ namespace Compression
...
+
+
+        //! Return true if there is an compression interface registered with the specified id
+        //! @param algorithmId CompressionAlgorithmId to determine if an compression interface is registered
+        //! @return bool indicating if there is an compression interface with the id registered
+        [[nodiscard]] virtual bool IsRegistered(CompressionAlgorithmId) const = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/CompressionInterfaceAPI.h:106]: Should match the doxygen entry for ```algorithmId```
```suggestion
        [[nodiscard]] virtual bool IsRegistered(CompressionAlgorithmId algorithmId ) const = 0;
```

File: Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h [hunk 7]
```diff
@@ -70,28 +70,45 @@ namespace Compression
...
 
-        //! Registers a decompression interface with the decompression factory.
+
+        //! Registers decompression interface and takes ownership of it if registration is successful
+        //! @param algorithmId Unique id to associate with decompression interface
+        //! @param decompressionInterface decompression interface to register
+        //! @return Success outcome if the decompression interface was successfully registered
+        //! Otherwise, a failure outcome with the decompression interface is forward back to the caller
+        virtual AZ::Outcome<void, AZStd::unique_ptr<IDecompressionInterface>> RegisterDecompressionInterface(
+            CompressionAlgorithmId algorithmId, AZStd::unique_ptr<IDecompressionInterface>) = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h:80]: Should match doxygen entry for ```decompressionInterface

File: Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h [hunk 8]
```diff
@@ -70,28 +70,45 @@ namespace Compression
...
-        //! and the unique_ptr refererence is unmodified and can be re-used by the caller
         //!
-        //! @param decompressionInterface unique pointer to decompression interface to register
-        //! @return true if the IDecompressionInterface was successfully registered, otherwise false
-        virtual bool RegisterDecompressionInterface(AZStd::unique_ptr<IDecompressionInterface>&&) = 0;
+        //! Registers decompression interface, but does not take ownership of it
+        //! @param algorithmId Unique id to associate with decompression interface
+        //! @param decompressionInterface decompression interface to register
+        //! @return true if the ICompressionInterface was successfully registered
+        virtual bool RegisterDecompressionInterface(CompressionAlgorithmId algorithmId, IDecompressionInterface&) = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h:91]: Should match doxygen entry for ```decompressionInterface```
```suggestion
        virtual bool RegisterDecompressionInterface(CompressionAlgorithmId algorithmId, IDecompressionInterface& decompressionInterface) = 0;
```

File: Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h [hunk 9]
```diff
@@ -70,28 +70,45 @@ namespace Compression
...
-        //! @return pointer to the DeCompression Interface or nullptr if not found
-        virtual IDecompressionInterface* FindDecompressionInterface(CompressionAlgorithmId) const = 0;
+        //! @param decompressionAlgorithmId unique Id that identifies the decompression interface
+        //! @return true if the unregistration is successful
+        virtual bool UnregisterDecompressionInterface(CompressionAlgorithmId algorithmId) = 0;
+
+        //! Queries the decompression interface with the decompression algorithmd Id
+        //! @param decompressionAlgorithmId unique Id of decompression interface to query
+        //! @return pointer to the decompression tnterface or nullptr if not found
+        [[nodiscard]] virtual IDecompressionInterface* FindDecompressionInterface(CompressionAlgorithmId) const = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h:102]: Should match doxygen entry for ```decompressionAlgorithmId```
```suggestion
        [[nodiscard]] virtual IDecompressionInterface* FindDecompressionInterface(CompressionAlgorithmId decompressionAlgorithmId) const = 0;
```

File: Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h [hunk 10]
```diff
@@ -70,28 +70,45 @@ namespace Compression
...
+
+
+        //! Return true if there is an decompression interface registered with the specified id
+        //! @param algorithmId CompressionAlgorithmId to determine if an decompression interface is registered
+        //! @return bool indicating if there is an decompression interface with the id registered
+        [[nodiscard]] virtual bool IsRegistered(CompressionAlgorithmId) const = 0;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Include/Compression/DecompressionInterfaceAPI.h:108]: Should match doxygen entry for ```algorithmId```
```suggestion
        [[nodiscard]] virtual bool IsRegistered(CompressionAlgorithmId algorithmId) const = 0;
```

File: Gems/Compression/Code/Source/Clients/DecompressionRegistrarImpl.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            DecompressionInterfacePtr m_decompressionInterface;
+        };
+
+        using IdToDecompressionInterfaceMap = AZStd::vector<DecompressionIdIndexEntry>;
+
+        //! Searches within the decompression interface array for the decompression interface registered with the specified id
+        //! @param algorithmId Unique Id of decompression interface to locate
+        //! @return iterator pointing the decompression interface registered with the specified CompressionAlgorithmId
+        //! NOTE: It is responsibility of the caller to lock the decompression interface mutex to protect the search
+        typename IdToDecompressionInterfaceMap::const_iterator FindDecompressionInterfaceImpl(CompressionAlgorithmId) const;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Source/Clients/DecompressionRegistrarImpl.h:75]: Should match doxygen entry for ```algorithmId```
```suggestion
        typename IdToDecompressionInterfaceMap::const_iterator FindDecompressionInterfaceImpl(CompressionAlgorithmId algorithmId) const;
```

File: Gems/Compression/Code/Source/Tools/CompressionRegistrarImpl.h [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            CompressionInterfacePtr m_compressionInterface;
+        };
+
+        using IdToCompressionInterfaceMap = AZStd::vector<CompressionIdIndexEntry>;
+
+        //! Searches within the compression interface array for the compression interface registered with the specified id
+        //! @param algorithmId Unique Id of compression interface to locate
+        //! @return iterator pointing the compression interface registered with the specified CompressionAlgorithmId
+        //! NOTE: It is responsibility of the caller to lock the compression interface mutex to protect the search
+        typename IdToCompressionInterfaceMap::const_iterator FindCompressionInterfaceImpl(CompressionAlgorithmId) const;
```
Comment (spham-amzn) [inline Gems/Compression/Code/Source/Tools/CompressionRegistrarImpl.h:75]: Should match the doxygen entry for ```algorithmId```
```suggestion
        typename IdToCompressionInterfaceMap::const_iterator FindCompressionInterfaceImpl(CompressionAlgorithmId algorithmId) const;
```

**Review by spham-amzn — APPROVED** [review]

**Pull request merged**